Signal/src/ViewControllers/ConversationView/MessagesViewController.m

```diff
@@ -166,6 +166,19 @@ @interface MessagesViewController () <AVAudioPlayerDelegate,
 @property (nonatomic) TSThread *thread;
 @property (nonatomic) TSMessageAdapter *lastDeliveredMessage;
 @property (nonatomic) YapDatabaseConnection *editingDatabaseConnection;
+
+// These two properties must be updated in lockstep.
+//
+// * The first (required) step is to update uiDatabaseConnection using beginLongLivedReadTransaction.
+// * The second (required) step is to update messageMappings.
+// * The third (optional) step is to update the messageMappings range using
+//   updateMessageMappingRangeOptions.
+// * The steps must be done in strict order.
+// * If we do any of the steps, we must do all of the required steps.
+// * We can't use messageMappings in between the first and second steps; e.g.
+//   we can't do any layout, since that uses numberOfItemsInSection: and
+//   interactionAtIndexPath: which use the messageMappings.
+// * If we do the third step, we must call resetContentAndLayout afterward.
```

Comment:
It took so much blood sweat and tears to learn this that I figured I would document it.